org.eclipse.xtext.xbase.tests/src/org/eclipse/xtext/xbase/tests/jvmmodel/JvmTypesBuilderTest.java

```diff
@@ -0,0 +1,447 @@
+	}
+
+	@Test
+	public void testSetBody_02() throws Exception {
+		XNullLiteral expr = XbaseFactory.eINSTANCE.createXNullLiteral();
+		XtextResource res = new XtextResource();
+		res.setLanguageName("org.eclipse.xtext.xbase.Xbase");
+		JvmOperation op = typesFactory.createJvmOperation();
+		jvmTypesBuilder.operator_add(res.getContents(), op);
+		jvmTypesBuilder.operator_add(res.getContents(), expr);
+		jvmTypesBuilder.setBody(op, (ITreeAppendable it) -> {
+			StringConcatenation builder = new StringConcatenation();
+			builder.append("bar");
+			it.append(builder);
+		});
+		Assert.assertEquals(1, op.eAdapters().size());
+		jvmTypesBuilder.setBody(op, expr);
+		Assert.assertEquals(op, containerProvider.getLogicalContainer(expr));
+		Assert.assertEquals(0, op.eAdapters().size());
+		jvmTypesBuilder.setBody(op, (ITreeAppendable it) -> {
+			new StringConcatenation().append("bar");
+			it.append(new StringConcatenation());
+		});
+		Assert.assertEquals(1, op.eAdapters().size());
+		Assert.assertNull(containerProvider.getLogicalContainer(expr));
+	}
+
+	@Test
+	public void testSetBody_03() throws Exception {
+		JvmOperation op = typesFactory.createJvmOperation();
+		jvmTypesBuilder.setBody(op, new StringConcatenationClient() {
+			@Override
+			protected void appendTo(StringConcatenationClient.TargetStringConcatenation builder) {
+				builder.append("foo");
+			}
+		});
+		jvmTypesBuilder.setBody(op, new StringConcatenationClient() {
+			@Override
+			protected void appendTo(StringConcatenationClient.TargetStringConcatenation builder) {
+				builder.append("bar");
+			}
+		});
+		Assert.assertEquals(1, op.eAdapters().size());
+	}
+
+	@Test
+	public void testSetBody_04() throws Exception {
+		XNullLiteral expr = XbaseFactory.eINSTANCE.createXNullLiteral();
+		XtextResource res = new XtextResource();
+		res.setLanguageName("org.eclipse.xtext.xbase.Xbase");
+		JvmOperation op = typesFactory.createJvmOperation();
+		jvmTypesBuilder.operator_add(res.getContents(), op);
+		jvmTypesBuilder.operator_add(res.getContents(), expr);
+		jvmTypesBuilder.setBody(op, new StringConcatenationClient() {
+			@Override
+			protected void appendTo(StringConcatenationClient.TargetStringConcatenation builder) {
+				builder.append("bar");
+			}
+		});
+		Assert.assertEquals(1, op.eAdapters().size());
+		jvmTypesBuilder.setBody(op, expr);
+		Assert.assertEquals(op, containerProvider.getLogicalContainer(expr));
+		Assert.assertEquals(0, op.eAdapters().size());
+		jvmTypesBuilder.setBody(op, new StringConcatenationClient() {
+			@Override
+			protected void appendTo(StringConcatenationClient.TargetStringConcatenation builder) {
+				builder.append("bar");
+			}
+		});
+		Assert.assertEquals(1, op.eAdapters().size());
+		Assert.assertNull(containerProvider.getLogicalContainer(expr));
+	}
+
+	@Test
+	@Ignore
+	public void testAddNull() throws Exception {
+		BasicEList<String> list = new BasicEList<String>();
+		Assert.assertTrue(list.isEmpty());
```

Comment:
The following comment is missing that explains why the test case is currently ignored: `//FIXME list += null`